### ISSUE TITLE
Make code more modular.

### ISSUE THREAD

**Issue by verdsveven** [Stargarth/Goblocks#1]
I suggest separating the necessary functions from the unnecessary ones. I.e. the built-ins such as date, cpu and memory should be separated from the rest (maybe put on a separate branch or patch). The RunCmd and Schedule functions should be part of the main package. FunctionMap should not be invoked in the main file either (it is unnecessary to call it when one for example does not want to use the builtin functions). This is all so as to make the program more suckless.

TLDR; The inclusion of the possibility to launch built-in functions is great, but should be made separate so as to better adhere to the suckless philosophy. If the user does not wish to utilise the built-in functions, he need not; if he wants to add more built-in functions, he should be able to.

Edit: it's ttecluk from reddit, we already talked about this.

**Comment by Ldaxar** [comment]
Hey @verdsveven !

I just made some changes to the repository. I moved essential functionality out of builtins to provide clearer separation. Now you can just delete builtins.go and change 1 line in main file to have a completely bare-bones experience :P. As to patching I'm not really fond of it when it comes to such a small project.

1. I consider built in programs to be part of core functionality as high performance is one of core objectives.
2. It's a small project and I don't have a website to post patches neither I can think of any easy(for user) and simple(for me) patch discovery process. I think it would cause more confusion rather than anything else.

With all that being said, I will make new functionality as separate programs with separate repos. They will be launched in "conventional" goblocks agnostic way. I'm working on weather forecasting program and it will just print to stdout, so it will be compatible with goblocks, dwmblocks, i3stuff etc. :D

**Comment by verdsveven** [comment]
The changes look good @Stargarth . This is a good step in making it more agnostic. I like the high performance ideas you are going for. My last issue is with the FunctionMap variable. Any way to make it automatically map it to the built in programs instead of manually mapping each time? For instance #Date turns into Date automatically. As far I understand the only purpose of FunctionMap is to do that, but manually.
Edit: I might be wrong about its purpose, please correct me if this is the case. Can't this be done in the config file? Add a "built-in" option to the struct; letting the program know that it is a builtin function?

**Comment by Ldaxar** [comment]
> Any way to make it automatically map it to the built in programs instead of manually mapping each time?  For instance #Date turns into Date automatically.
>
Yeah, that's a tough one - I will think about it.
>As far I understand the only purpose of FunctionMap is to do that, but manually.
>
Not really, it's done that way to run arbitrary functions based on arbitrary string input through the fact that function are first class citizens in Go. In Go and all compiled languages you can't treat input as a code (without some weird reflection shenanigans that is). Languages that can do that are called interpreted languages. Using those you can just append brackets () to input that you receive and it will run as a function.

Hope that this makes some sense :P

**Comment by verdsveven** [comment]
> Not really, it's done that way to run arbitrary functions based on arbitrary string input through the fact that function are first class citizens in Go. In Go and all compiled languages you can't treat input as a code (without some weird reflection shenanigans that is). Languages that can do that are called interpreted languages. Using those you can just append brackets () to input that you receive and it will run as a function.

Yeah, I get it. It's really a nitpick. Maybe you could cycle through all instances including # in the config and relate to them into the map by using the append() function?

**Comment by Ldaxar** [comment]
> Maybe you could cycle through all instances including # in the config and relate to them into the map by using the append() function?
>
Can you provide an example?

**Comment by Ldaxar** [comment]
I didn't manage to figure out automatic mapping and there have been no updates for over a week.
I'm closing down this issue.

**Comment by verdsveven** [comment]
> I didn't manage to figure out automatic mapping and there has been no updates for over a week.
> I'm closing down this issue.

I was busy recently and didn't manage to look into it. However, it seems that you are right @Stargarth . There seems to be no way of doing automatic mapping.
